src/zcbenchmarks.cpp

```diff
@@ -117,35 +119,39 @@ double benchmark_solve_equihash(bool time)
                                     nonce.begin(),
                                     nonce.size());
 
-    if (time)
-        timer_start();
+    struct timeval tv_start;
+    timer_start(tv_start);
     std::set<std::vector<unsigned int>> solns;
     EhOptimisedSolveUncancellable(n, k, eh_state,
                                   [](std::vector<eh_index> soln) { return false; });
-    if (time)
-        return timer_stop();
-    else
-        return 0;
+    return timer_stop(tv_start);
 }
 
-double benchmark_solve_equihash_threaded(int nThreads)
+std::vector<double> benchmark_solve_equihash_threaded(int nThreads)
 {
-    boost::thread_group solverThreads;
-    timer_start();
-    for (int i = 0; i < nThreads; i++)
-        solverThreads.create_thread(boost::bind(&benchmark_solve_equihash, false));
-    solverThreads.join_all();
-    return timer_stop();
+    std::vector<double> ret;
+    std::vector<boost::shared_future<double>> tasks;
+    for (int i = 0; i < nThreads; i++) {
+        tasks.emplace_back(boost::async(&benchmark_solve_equihash));
+    }
+    auto it = boost::wait_for_any(tasks.begin(),tasks.end());
+    while(it != tasks.end()) {
+        ret.push_back(it->get());
+        tasks.erase(it);
+        it = boost::wait_for_any(tasks.begin(),tasks.end());
```

Comment:
This is a slightly strange API (in what it uses iterators for, essentially as just a kind of pointer), but I looked at [its documentation](http://www.boost.org/doc/libs/1_46_0/doc/html/thread/synchronization.html#thread.synchronization.futures.reference.wait_for_any) and I guess this usage is correct.